components/service/glean/src/main/java/mozilla/components/service/glean/histogram/FunctionalHistogram.kt

```diff
@@ -0,0 +1,185 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+package mozilla.components.service.glean.histogram
+
+import mozilla.components.support.ktx.android.org.json.tryGetLong
+import org.json.JSONObject
+import java.lang.Math.pow
+import kotlin.math.log
+
+/**
+ * This class represents a histogram where the bucketing is performed by a
+ * function, rather than pre-computed buckets. It is meant to help serialize
+ * and deserialize data to the correct format for transport and storage, as well
+ * as performing the calculations to determine the correct bucket for each sample.
+ *
+ * The bucket index of a given sample is determined with the following function:
+ *
+ *     i = ⌊n log₂(𝑥)⌋
+ *
+ * In other words, there are n buckets for each power of 2 magnitude.
+ *
+ * The value of 8 for n was determined experimentally based on existing data to have sufficient
+ * resolution.  See https://bugzilla.mozilla.org/show_bug.cgi?id=1565253
+ *
+ * @param values a map containing the minimum bucket value mapped to the accumulated count
+ * @param sum the accumulated sum of all the samples in the histogram
+ */
+data class FunctionalHistogram(
+    // map from bucket limits to accumulated values
+    val values: MutableMap<Long, Long> = mutableMapOf(),
+    var sum: Long = 0
+) {
+    companion object {
+        // The base of the logarithm used to determine bucketing
+        internal const val LOG_BASE = 2.0
+
+        // The buckets per each order of magnitude of the logarithm.
+        internal const val BUCKETS_PER_MAGNITUDE = 8.0
```

Comment:
Does this need to be configurable?

Reply:
Yeah, I guess it does.  I've updated this PR so it's a parameter...